File: Activities/Activity 4/Gearbox.cpp
#include <iostream>
#include <cmath>
#include <limits>

using namespace std;

int main() {

    double maximumSpeed, minimumSpeed;
    cout << "Enter a minimum and maximum speed (in revolutions per minute): ";
    cin >> minimumSpeed >> maximumSpeed;

    while(cin.fail() || minimumSpeed <= 0 || maximumSpeed <= 0) {
        if(cin.fail()) {
            cout << "Your input was invalid. Please try again." << endl;
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
        } else {
            cout << "You must enter two positive number values. Please try again." << endl;
        }

        cout << "Enter a minimum and maximum speed (in revolutions per minute): ";
        cin >> minimumSpeed >> maximumSpeed;
    }

    if(maximumSpeed < minimumSpeed) {
        double temp = maximumSpeed;
        maximumSpeed = minimumSpeed;
        minimumSpeed = temp;
    }

    double speedRatio = pow(maximumSpeed / minimumSpeed, 0.2);
    cout << "The ratio between successive speeds of a six-speed gearbox"
        << "\nwith a minimum speed of " << minimumSpeed
        << " rpm and a maximum speed of " << maximumSpeed
        << " rpm is " << speedRatio << endl;

    for(int i = 0; i < 6; i++)
        cout << "Speed " << i + 1 << " is " << minimumSpeed * pow(speedRatio, i) << " rpm" << endl;

}

File: Activities/Activity 4/Gearbox.java
import java.util.InputMismatchException;
import java.util.Scanner;

public class Gearbox {

	public static void main(String[] args) {
		
		Scanner s = new Scanner(System.in);
		double minimumSpeed = 0, maximumSpeed = 0;
		{
			boolean valid;
			do {
				try {
					System.out.print("Enter a minimum and maximum speed (in revolutions per minute): ");
					minimumSpeed = s.nextDouble();
					maximumSpeed = s.nextDouble();
					
					if(minimumSpeed <= 0 || maximumSpeed <= 0) {
						System.out.println("You must enter two positive number values. Please try again.");
						valid = false;
					}
					else valid = true;
				} catch(InputMismatchException ime) {
					System.out.println("Your input was invalid. Please try again.");
					s.nextLine();
					valid = false;
				}
			} while(!valid);
		}
		s.close();

	    if(maximumSpeed < minimumSpeed) {
	        double temp = maximumSpeed;
	        maximumSpeed = minimumSpeed;
	        minimumSpeed = temp;
	    }

	    double speedRatio = Math.pow(maximumSpeed / minimumSpeed, 0.2);
	    System.out.println("The ratio between successive speeds of a six-speed gearbox"
	        + "\nwith a minimum speed of " + minimumSpeed
	        + " rpm and a maximum speed of " + maximumSpeed
	        + " rpm is " + speedRatio);

	    for(int i = 0; i < 6; i++)
	        System.out.println("Speed " + (i + 1) + " is " + minimumSpeed * Math.pow(speedRatio, i) + " rpm");

	}
	
}
